bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/firmware/Firmware.java

```diff
@@ -67,150 +57,89 @@
  * device is required afterwards) so that {@link FirmwareUpdateHandler}s can handle this information accordingly.
  *
  * @author Thomas Höfer - Initial contribution
+ * @author Dimitar Ivanov - FirmwareUID is replaced by ThingTypeUID and firmware version
  */
-public final class Firmware implements Comparable<Firmware> {
+public interface Firmware extends Comparable<Firmware> {
 
     /** The key for the requires a factory reset property. */
     public static final String PROPERTY_REQUIRES_FACTORY_RESET = "requiresFactoryReset";
 
-    private static final Logger LOGGER = LoggerFactory.getLogger(Firmware.class);
-
-    private final FirmwareUID uid;
-    private final String vendor;
-    private final String model;
-    private final boolean modelRestricted;
-    private final String description;
-    private final String version;
-    private final String prerequisiteVersion;
-    private final String changelog;
-    private final URL onlineChangelog;
-    private final transient InputStream inputStream;
-    private final String md5Hash;
-    private final Map<String, String> properties;
-
-    private transient byte[] bytes;
-
-    private final Version internalVersion;
-    private final Version internalPrerequisiteVersion;
-
-    private Firmware(Builder builder) {
-        this.uid = builder.uid;
-        this.version = builder.uid.getFirmwareVersion();
-        this.vendor = builder.vendor;
-        this.model = builder.model;
-        this.modelRestricted = builder.modelRestricted;
-        this.description = builder.description;
-        this.prerequisiteVersion = builder.prerequisiteVersion;
-        this.changelog = builder.changelog;
-        this.onlineChangelog = builder.onlineChangelog;
-        this.inputStream = builder.inputStream;
-        this.md5Hash = builder.md5Hash;
-        this.properties = Collections
-                .unmodifiableMap(builder.properties != null ? builder.properties : Collections.emptyMap());
-
-        this.internalVersion = new Version(this.version);
-        this.internalPrerequisiteVersion = this.prerequisiteVersion != null ? new Version(this.prerequisiteVersion)
-                : null;
-    }
-
     /**
-     * Returns the UID of the firmware.
+     * Returns the thing type UID, that this firmware is associated with.
      *
-     * @return the UID of the firmware (not null)
+     * @return the thing type UID
```

Comment:
i guess this is never null, isnt it?

Reply:
Yes, it is not null. I will apply the @NonNullByDefault annotations according  to the coding guidelines.